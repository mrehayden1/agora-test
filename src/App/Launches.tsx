import Launch from './Launch';
import * as L from './Launch';
import { Sort } from './Filters';

export type LaunchData = L.LaunchData;

type Props = {
  launches: null | LaunchData[];
  search: string;
  sort: Sort;
}

export default function Launches(props: Props) {

  const filteredAndSortedLaunches = props.launches && props.launches
    .filter((launch) => (
      launch
        .mission_name
        .toLowerCase()
        .indexOf(props.search.toLowerCase()) >= 0
    ))
    .sort((a, b) => {
      switch (props.sort) {
        case 'newest':
          return b.launch_date_unix - a.launch_date_unix;
        case 'oldest':
          return a.launch_date_unix - b.launch_date_unix;
        case 'name':
          return a.mission_name.localeCompare(b.mission_name);
        case 'name-desc':
          return b.mission_name.localeCompare(a.mission_name);
        default:
          throw Error('Unreachable code.');
      }
    });

  return (
    filteredAndSortedLaunches && (
      <>
        {filteredAndSortedLaunches.map((launch, i) => (
          <Launch key={i} {...launch} />
        ))}
      </>
    )
  );
}
