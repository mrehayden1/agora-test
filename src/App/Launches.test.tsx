import { shallow } from 'enzyme';
import React from 'react';

import Launches, { LaunchData } from './Launches';

const launches: LaunchData[] = [{
  details: 'details one',
  launch_date_unix: 1631655161,
  mission_name: 'mission one',
  rocket: {
    rocket_name: 'rocket one',
    rocket_type: 'rocket type one'
  }
}, {
  details: 'details two',
  launch_date_unix: 1631655160,
  mission_name: 'mission two',
  rocket: {
    rocket_name: 'rocket two',
    rocket_type: 'rocket type two'
  }
}];

test('renders launches in correct order', () => {
  const component = shallow(<Launches launches={launches} sort={'newest'} search={''} />);
  expect(component).toMatchSnapshot();
});

test('renders launches in correct order (oldest)', () => {
  const component = shallow(<Launches launches={launches} sort={'oldest'} search={''} />);
  expect(component).toMatchSnapshot();
});

test('renders launches in correct order (name)', () => {
  const component = shallow(<Launches launches={launches} sort={'name'} search={''} />);
  expect(component).toMatchSnapshot();
});

test('renders launches in correct order (name desc)', () => {
  const component = shallow(<Launches launches={launches} sort={'name-desc'} search={''} />);
  expect(component).toMatchSnapshot();
});

test('renders launches when search present', () => {
  const component = shallow(<Launches launches={launches} sort={'newest'} search={'one'} />);
  expect(component).toMatchSnapshot();
});
