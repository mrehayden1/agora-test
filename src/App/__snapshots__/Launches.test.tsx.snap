// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders launches in correct order (name desc) 1`] = `
<Fragment>
  <Launch
    details="details two"
    key="0"
    launch_date_unix={1631655160}
    mission_name="mission two"
    rocket={
      Object {
        "rocket_name": "rocket two",
        "rocket_type": "rocket type two",
      }
    }
  />
  <Launch
    details="details one"
    key="1"
    launch_date_unix={1631655161}
    mission_name="mission one"
    rocket={
      Object {
        "rocket_name": "rocket one",
        "rocket_type": "rocket type one",
      }
    }
  />
</Fragment>
`;

exports[`renders launches in correct order (name) 1`] = `
<Fragment>
  <Launch
    details="details one"
    key="0"
    launch_date_unix={1631655161}
    mission_name="mission one"
    rocket={
      Object {
        "rocket_name": "rocket one",
        "rocket_type": "rocket type one",
      }
    }
  />
  <Launch
    details="details two"
    key="1"
    launch_date_unix={1631655160}
    mission_name="mission two"
    rocket={
      Object {
        "rocket_name": "rocket two",
        "rocket_type": "rocket type two",
      }
    }
  />
</Fragment>
`;

exports[`renders launches in correct order (oldest) 1`] = `
<Fragment>
  <Launch
    details="details two"
    key="0"
    launch_date_unix={1631655160}
    mission_name="mission two"
    rocket={
      Object {
        "rocket_name": "rocket two",
        "rocket_type": "rocket type two",
      }
    }
  />
  <Launch
    details="details one"
    key="1"
    launch_date_unix={1631655161}
    mission_name="mission one"
    rocket={
      Object {
        "rocket_name": "rocket one",
        "rocket_type": "rocket type one",
      }
    }
  />
</Fragment>
`;

exports[`renders launches in correct order 1`] = `
<Fragment>
  <Launch
    details="details one"
    key="0"
    launch_date_unix={1631655161}
    mission_name="mission one"
    rocket={
      Object {
        "rocket_name": "rocket one",
        "rocket_type": "rocket type one",
      }
    }
  />
  <Launch
    details="details two"
    key="1"
    launch_date_unix={1631655160}
    mission_name="mission two"
    rocket={
      Object {
        "rocket_name": "rocket two",
        "rocket_type": "rocket type two",
      }
    }
  />
</Fragment>
`;

exports[`renders launches when search present 1`] = `
<Fragment>
  <Launch
    details="details one"
    key="0"
    launch_date_unix={1631655161}
    mission_name="mission one"
    rocket={
      Object {
        "rocket_name": "rocket one",
        "rocket_type": "rocket type one",
      }
    }
  />
</Fragment>
`;
